src/libvmmalloc/libvmmalloc.c

```diff
@@ -371,17 +391,199 @@ libvmmalloc_create(const char *dir, size_t size)
 }
 
 /*
+ * libvmmalloc_clone - (internal) clone the entire pool
+ */
+static void *
+libvmmalloc_clone(void)
+{
+	LOG(3, NULL);
+
+	Fd_clone = util_tmpfile(Dir, Vmp->size);
+	if (Fd_clone == -1)
+		return NULL;
+
+	void *addr = mmap(NULL, Vmp->size, PROT_READ|PROT_WRITE,
+			MAP_SHARED, Fd_clone, 0);
+	if (addr == MAP_FAILED) {
+		LOG(1, "!mmap");
+		(void) close(Fd_clone);
+		return NULL;
+	}
+
+	LOG(3, "copy the entire pool file: dst %p src %p size %zu",
+			addr, Vmp->addr, Vmp->size);
+
+	util_range_rw(Vmp->addr, sizeof (struct pool_hdr));
+	memcpy(addr, Vmp->addr, Vmp->size);
+	util_range_none(Vmp->addr, sizeof (struct pool_hdr));
+
+	return addr;
+}
+
+/*
+ * libvmmalloc_prefork -- (internal) prepare for fork()
+ *
+ * Clones the entire pool or remaps it with MAP_PRIVATE flag.
+ */
+static void
+libvmmalloc_prefork(void)
+{
+	LOG(3, NULL);
+
+	/*
+	 * There's no need to grab any locks here, as jemalloc pre-fork handler
+	 * is executed first, and it does all the synchronization.
+	 */
+
+	ASSERTne(Vmp, NULL);
+	ASSERTne(Dir, NULL);
+
+	void *addr = Vmp->addr;
+	size_t size = Vmp->size;
+
+	if (Private) {
+		LOG(3, "already mapped as private - do nothing");
+		return;
+	}
+
+	switch (Forkopt) {
+	case 3:
+		/* clone the entire pool; if it fails - remap it as private */
+		LOG(3, "clone or remap");
+
+	case 2:
+		LOG(3, "clone the entire pool file");
+
+		if (libvmmalloc_clone() != NULL)
+			break;
+
+		if (Forkopt == 2) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"pool cloning failed\n");
+			exit(1);
```

Comment:
Ha! I planned to change each `exit(1)` to `abort()`, but this affects a couple of negative unit tests, so I left it for another patch.
Also, I didn't want to use `out_fatal()`, as it logs only if the log level is >=1 (!!!).